sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
+    size_t LID = NDId.get_local_linear_id();
+    LocalReds[LID] = Reducer.MValue;
+    LocalReds[WGSize] = Identity;
+    DoReducePartialSumsInLastWG[0] = 0;
+    sycl::detail::workGroupBarrier();
+
+    // Tree-reduction: reduce the local array LocalReds[:] to LocalReds[0].
+    // LocalReds[WGSize] accumulates last/odd elements when the step
+    // of tree-reduction loop is not even.
+    size_t PrevStep = WGSize;
+    for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+      if (LID < CurStep)
+        LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+      else if (LID == CurStep && (PrevStep & 0x1))
+        LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+      sycl::detail::workGroupBarrier();
+      PrevStep = CurStep;
+    }
+
+    if (LID == 0) {
+      auto V = BOp(LocalReds[0], LocalReds[WGSize]);
+      if (NWorkGroups == 1 && IsUpdateOfUserVar)
+        V = BOp(V, *Reduction::getOutPointer(Out));
+      // if NWorkGroups == 1, then PartialsSum and Out point to same memory.
+      Reduction::getOutPointer(PartialSums)[NDId.get_group_linear_id()] = V;
+
+      auto NFinished =
+          atomic_ref<int, memory_order::relaxed, memory_scope::device,
+                     access::address_space::global_space>(
+              NWorkGroupsFinished[0]);
+      if (++NFinished == NWorkGroups && NWorkGroups > 1)
+        DoReducePartialSumsInLastWG[0] = 1;
+    }
+
+    sycl::detail::workGroupBarrier();
+    if (DoReducePartialSumsInLastWG[0]) {
+      auto LocalSum = Identity;
+      for (size_t I = LID; I < NWorkGroups; I += WGSize)
+        LocalSum = BOp(LocalSum, Reduction::getOutPointer(PartialSums)[I]);
+      LocalReds[LID] = LocalSum;
+      LocalReds[WGSize] = Identity;
+      sycl::detail::workGroupBarrier();
+
+      size_t PrevStep = WGSize;
+      for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+        if (LID < CurStep)
+          LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+        else if (LID == CurStep && (PrevStep & 0x1))
+          LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+        sycl::detail::workGroupBarrier();
+        PrevStep = CurStep;
+      }
+      if (LID == 0) {
+        auto V = BOp(LocalReds[0], LocalReds[WGSize]);
+        if (IsUpdateOfUserVar)
+          V = BOp(V, *Reduction::getOutPointer(Out));
+        Reduction::getOutPointer(Out)[0] = V;
+      }
+    }
+  });
+}
+
+template <typename KernelName, typename KernelType, int Dims, class Reduction>
+void reduCGFunc(handler &CGH, KernelType KernelFunc, const range<Dims> &Range,
+                size_t MaxWGSize, uint32_t NumEUThreads, Reduction &Redu) {
+  size_t NWorkItems = Range.size();
+  size_t WGSize = std::min(NWorkItems, MaxWGSize);
+  size_t NWorkGroups = NWorkItems / WGSize;
+  if (NWorkItems % WGSize)
+    NWorkGroups++;
+  size_t MaxNWorkGroups = NumEUThreads;
```

Comment:
The mapping to Intel GPUs is such that 1 EU thread == 1 sub-group.  It's not clear to me that setting the number of work groups equal to the number of EU threads is particularly meaningful if the work-group size is large.

Reply:
This perhaps can be tuned additionally later. I see now that these heuristics give the best results. Lowering number of work-groups or reducing the size of work-groups gives slower perf.

Reply:
Apologies if this comment appears twice -- having some GitHub problems.  I think we should add an explicit TODO in the implementation of MaxNumConcurrentWorkGroups saying that it needs to be tuned for other devices.

Reply:
Thank you for the comment. Heuristics definitely need some additional tuning.
There is a comment saying exactly that: https://github.com/v-klochkov/llvm/blob/public_vklochkov_reduction_range_review/sycl/source/detail/reduction.cpp#L57

I'll add "TODO: " to it in a separate [NFC] patch

Reply:
Yeah, I tried to post this message directly tied to the comment but GitHub wouldn't let me do it!  Adding the TODO in a separate NFC patch sounds good to me.

Reply:
I added the TODO comment to this PR: https://github.com/intel/llvm/pull/4361 (commit: 900de465b5b1a281ef65375d6289534e93fb6072)

Reply:
I added the TODO comment to this PR: https://github.com/intel/llvm/pull/4361 (commit: 900de465b5b1a281ef65375d6289534e93fb6072)